underfs/s3a/src/main/java/alluxio/underfs/s3a/S3AInputStream.java

```diff
@@ -125,29 +133,54 @@ public long skip(long n) throws IOException {
     if (n <= 0) {
       return 0;
     }
-    closeStream();
     mPos += n;
-    openStream();
     return n;
   }
 
   /**
-   * Opens a new stream at mPos if the wrapped stream mIn is null.
+   * Seek if we could reuse the existing stream or create a new one.
+   * @param length the content length to be read
    */
-  private void openStream() throws IOException {
-    if (mIn != null) { // stream is already open
+  private void seekStream(int length) throws IOException {
+    if (mIn != null && mPos == mInPos) { // stream is already open and at the expected position
       return;
     }
+    if (mIn != null && mPos > mInPos) { // stream is already open but the caller is seeking forward
+      long skip = mPos - mInPos;
+      if (skip < mIn.available()) {
```

Comment:
IIRC, `available` is not supported on all input streams, and [may just return 0](https://docs.oracle.com/javase/8/docs/api/java/io/InputStream.html#available--). Do you know if this method is implemented properly for the underlying stream?

Reply:
Good question!  I believed so, I will double check and get back to you later.